pkg/controller/upgradeconfig/cluster_upgrader.go

```diff
@@ -0,0 +1,737 @@
+		log.Error(err, "failed to get original machinesets")
+		return false, err
+	}
+	if len(originalMachineSets.Items) == 0 {
+		log.Info("failed to get machineset")
+		return false, fmt.Errorf("failed to get original machineset")
+	}
+
+	updated := false
+	for _, ms := range originalMachineSets.Items {
+
+		found := false
+		for _, ums := range upgradeMachinesets.Items {
+			if ums.Name == ms.Name+"-upgrade" {
+				found = true
+			}
+		}
+		if found {
+			log.Info(fmt.Sprintf("machineset for upgrade already created :%s", ms.Name))
+			continue
+		}
+		updated = true
+		replica := int32(1)
+		newMs := ms.DeepCopy()
+
+		newMs.ObjectMeta = metav1.ObjectMeta{
+			Name:      ms.Name + "-upgrade",
+			Namespace: ms.Namespace,
+			Labels: map[string]string{
+				LABEL_UPGRADE: "true",
+			},
+		}
+		newMs.Spec.Replicas = &replica
+		newMs.Spec.Template.Labels[LABEL_UPGRADE] = "true"
+		newMs.Spec.Selector.MatchLabels[LABEL_UPGRADE] = "true"
+		log.Info(fmt.Sprintf("creating machineset %s for upgrade", newMs.Name))
+
+		err = c.Create(context.TODO(), newMs)
+		if err != nil {
+			log.Error(err, "failed to create machineset")
+			return false, err
+		}
+
+	}
+	if updated {
+		// New machineset created, machines must not ready at the moment, so skip following steps
+		return false, nil
+	}
+	nodes := &corev1.NodeList{}
+	err = c.List(context.TODO(), nodes)
+	if err != nil {
+		log.Error(err, "failed to list nodes")
+		return false, err
+	}
+	allNodeReady := true
+	for _, ms := range upgradeMachinesets.Items {
+		//We assume the create time is the start time for scale up extra compute nodes
+		startTime := ms.CreationTimestamp
+		if ms.Status.Replicas != ms.Status.ReadyReplicas {
+
+			if time.Now().After(startTime.Time.Add(TIMEOUT_SCALE_EXTRAL_NODES)) {
+				//TODO send out timeout alerts
+				log.Info(fmt.Sprintf("machineset provisioning timout"))
+			}
+			log.Info(fmt.Sprintf("not all mahines are ready for machineset:%s", ms.Name))
+			return false, nil
+		}
+		machines := &machineapi.MachineList{}
+		err := c.List(context.TODO(), machines, []client.ListOption{
+			client.InNamespace("openshift-machine-api"),
+			client.MatchingLabels{LABEL_UPGRADE: "true"},
+		}...)
+		if err != nil {
+			log.Error(err, "failed to list extra upgrade machine")
+			return false, err
+		}
+		nodeReady := false
+		var nodeName string
+		for _, node := range nodes.Items {
+			if node.Annotations["machine.openshift.io/machine"] == "openshift-machine-api/"+machines.Items[0].Name {
+				for _, con := range node.Status.Conditions {
+					if con.Type == corev1.NodeReady && con.Status == corev1.ConditionTrue {
+						nodeReady = true
+						nodeName = node.Name
+					}
+				}
+
+			}
+
+		}
+		if !nodeReady {
+			allNodeReady = false
+			if time.Now().After(startTime.Time.Add(TIMEOUT_SCALE_EXTRAL_NODES)) {
+				log.Info("node is not ready within 30mins")
+				//TODO send out timeout alerts
+				return false, fmt.Errorf("timeout waiting node:%s become ready", nodeName)
```

Comment:
Fixing small grammatical issues
```suggestion
				return false, fmt.Errorf("timeout waiting for node:%s to become ready", nodeName)
```